src/transport/emitter.ts

```diff
@@ -58,3 +59,66 @@ export function emitterFor(fn: TransportFunction, options = { binding: HTTP, mod
     }
   };
 }
+
+/**
+ * A static class to emit CloudEvents within an application
+ */
+export class Emitter extends EventEmitter {
+  /**
+   * Singleton store
+   */
+  static instance: Emitter | undefined = undefined;
+
+  /**
+   * Create an Emitter
+   * On v4.0.0 this class will only remains as Singleton to allow using the
+   * EventEmitter of NodeJS
+   */
+  private constructor() {
+    super();
+  }
+
+  /**
+   * Return or create the Emitter singleton
+   *
+   * @return {Emitter} return Emitter singleton
+   */
+  static getInstance(): Emitter {
+    if (!Emitter.instance) {
+      Emitter.instance = new Emitter();
+    }
+    return Emitter.instance;
+  }
+
+  /**
+   * Add a listener for eventing
+   *
+   * @param {string} event type to listen to
+   * @param {Function} listener to call on event
+   * @return {void}
+   */
+  static on(event: "cloudevent" | "newListener" | "removeListener", listener: (...args: any[]) => void): void {
```

Comment:
Why are we creating an event listening system for an event emitter?

I do not expect an eventing system like this.

Reply:
This is a shortcut to avoid the `getSingleton` that you did not like, I started to reimplement the EventEmitter interface, but found it overkill, as there are several other methods that can be useful. So instead of reinventing the wheel, I just added a shortcut method to streamline the process

Reply:
In my opinion, [the example added to the README](https://github.com/cloudevents/sdk-javascript/pull/366/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R96-R114) is very illustrative. This PR makes it so that you can set up your emitter function using `emitterFor()`, then have that respond to all `emit` events in the Node event system by actually sending the `CloudEvent`. To me that seems quite valuable. Imagine that you have a large system that is pushing out `CloudEvents`, perhaps even to multiple different endpoints. You could add multiple emitter functions at startup that each send to a different endpoint. Then by simply calling `new CloudEvent({...}).emit();` the `CloudEvents` is sent - with that single call - to all endpoints. I like it a lot.

Reply:
Yes this was the exact purpose, thanks for summarizing it so well

Reply:
Yup, the README example is great.

```js
const axios = require("axios").default;
const { emitterFor, Mode, CloudEvent, Emitter } = require("cloudevents");
function sendWithAxios(message) {
  // Do what you need with the message headers
  // and body in this function, then send the
  // event
  axios({
    method: "post",
    url: "...",
    data: message.body,
    headers: message.headers,
  });
}
const emit = emitterFor(sendWithAxios, { mode: Mode.BINARY });
// Set the emit
Emitter.on("cloudevent", emit);
...
// In any part of the code will send the event
new CloudEvent({ type, source, data }).emit();
```

I think the fundamental question is why are we adding back the `emit` function? And adding a `ensureDelivery` param?

A CloudEvent object and a CloudEvent emitter/receiver seem like separate concepts.

```js
const axios = require("axios").default;
const { emitterFor, Mode, CloudEvent, Emitter } = require("cloudevents");
function sendWithAxios(message) {
  // Do what you need with the message headers
  // and body in this function, then send the
  // event
  axios({
    method: "post",
    url: "...",
    data: message.body,
    headers: message.headers,
  });
}
const emit = emitterFor(sendWithAxios, { mode: Mode.BINARY });

// In any part of the code will send the event
const ce = new CloudEvent({ type, source, data });
emit(ce);
```

Folks can add an eventing layer on-top of the primitives here. Maybe point out an error in the above alternative code?

Reply:
The emitEvent function was adding because EventEmitter is not able to "wait" for all listeners to finish async jobs.
You might by design consider that you need to fail the whole process if you were not able to send the event, or you can decide to let it go depending on how critical to your business the CloudEvent is.
This is the SDK so I think it should be able to answer both needs instead of limiting to one interpretation.

Reply:
> So I wouldn't imagine a CloudEvent has a send/receive function.

Just to be clear, we are talking about two separate concepts. It's easy to get things mixed up when all of the terms are so similar. The two separate things as I see it are:

1. Sending a `CloudEvent` over some network transport using the `emitterFor()` factory to create a user defined function that is capable of converting a `CloudEvent` into a `Message` and sending that over the wire with a network transport (typically HTTP).
1. Using the Node.js built in `EventEmitter` capabilities to notify a [`listener`](https://nodejs.org/api/events.html#events_emitter_on_eventname_listener) that a `CloudEvent` should be sent using the method described in 1) point above. The `listener` in this case is the transport function created with `emitterFor()`. The new `CloudEvent#emit()` function just facilitates sending these using the `Emitter.emitEvent()` function.

This pull request is about using what is described in 2) to facilitate and orchestrate 1) in an efficient way across a large system that may need to send `CloudEvents` over network transport to multiple endpoints and do so efficiently by making use of built in Node.js `EventEmitter` capabilities.

> Maybe point out an error in the above alternative code?

This pull request is not an alternative to that code. It builds upon that code to add a very useful feature, in my opinion, with very low overhead using Node.js built in, idiomatic capabilities.